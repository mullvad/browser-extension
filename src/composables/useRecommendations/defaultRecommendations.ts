import { Recommendation } from './Recommendation.types';

export const defaultExtensions: Recommendation[] = [
  {
    type: 'extension',
    id: 'uBlock0@raymondhill.net',
    name: 'Install uBlock Origin',
    description: `uBlock Origin is not just a free and open-source “ad blocker“, it's a very efficient content blocker consuming minimal resources.`,
    homeUrl: 'https://ublockorigin.com/',
    icon: 'ubo64.png',
    enabled: false,
    installed: false,
    activated: false,
    ignored: false,
    ctaLabel: 'install',
    ctaUrl: 'https://addons.mozilla.org/firefox/addon/ublock-origin/',
  },
];

export const defaultSettings: Recommendation[] = [
  {
    type: 'setting',
    id: 'https-only-mode',
    name: 'Use HTTPS-only mode',
    description:
      'Enabling HTTPS-only mode makes sure all of your connections are encrypted with HTTPS, and warn you if only HTTP unsafe mode is available.',
    homeUrl:
      'https://support.mozilla.org/en-US/kb/https-only-prefs#w_enabledisable-https-only-mode',
    activated: false,
    ignored: false,
    ctaLabel: undefined,
  },
  {
    type: 'setting',
    id: 'default-search',
    name: 'Change your default search engine',
    description:
      'Your default search engine is not respectful of your privacy. We recommend you switch to a more private one.',
    homeUrl:
      'https://mullvad.net/en/blog/2021/2/24/dont-duck-the-issue-consider-your-privacy-and-search-engines/',
    activated: false,
    ignored: false,
    ctaLabel: undefined,
  },
];

export const defaultRecommendations = [...defaultExtensions, ...defaultSettings];

export const isRecommended = (id: string) => {
  return defaultRecommendationsIds.includes(id);
};

export const defaultRecommendationsIds = defaultRecommendations.map((rec) => rec.id);
