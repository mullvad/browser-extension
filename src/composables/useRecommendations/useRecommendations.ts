import { computed, watch } from 'vue';
import { management } from 'webextension-polyfill';

import useBrowserStorageLocal from '@/composables/useBrowserStorageLocal';
import useHttpsOnly from '@/composables/useHttpsOnly';
import { Recommendation } from '@/composables/useRecommendations/Recommendation.types';
import {
  defaultRecommendations,
  defaultRecommendationsIds,
} from '@/composables/useRecommendations/defaultRecommendations';

const recommendations = useBrowserStorageLocal<Recommendation[]>(
  'recommendations',
  defaultRecommendations,
);

const updateRecommendation = (id: string, modification: Partial<Recommendation>) => {
  recommendations.value = recommendations.value.map((recommendation) =>
    recommendation.id === id ? { ...recommendation, ...modification } : recommendation,
  );
};

const cleanOutdated = () => {
  recommendations.value = recommendations.value.filter(
    (recommendation) => !['Privacy Badger', 'Cookie AutoDelete'].includes(recommendation.name),
  );
};

const getRecommendationById = (id: string) => {
  return recommendations.value.find((rec) => rec.id === id);
};

const updateExtensions = async () => {
  const installedAddons = (await management.getAll()).filter((extension) =>
    defaultRecommendationsIds.includes(extension.id),
  );

  const enabledExtensionIds = installedAddons
    .filter((addon) => addon.enabled)
    .map((addons) => addons.id);

  installedAddons.forEach((extension) => {
    const enabled = enabledExtensionIds.includes(extension.id);

    const partialUpdate: Partial<Recommendation> = {
      ctaLabel: enabled ? undefined : 'enable',
      enabled,
      installed: true,
      activated: enabled,
    };

    updateRecommendation(extension.id, partialUpdate);
  });
};

const updateHttpsOnly = async () => {
  const httpsOnly = await useHttpsOnly();

  updateRecommendation('https-only-mode', { activated: httpsOnly });
};

const updateDefaultSearch = async () => {
  const searchEngines = await browser.search.get();

  const defaultSearchEngine = searchEngines
    .filter((search) => search.isDefault === true)[0]
    .name.toLowerCase();

  const isNotPrivacyFriendly = ['google', 'bing', 'yahoo', 'yandex', 'baidu', 'amazon'].some(
    (search) => defaultSearchEngine.includes(search),
  );

  updateRecommendation('default-search', { activated: !isNotPrivacyFriendly });
};

const updateSettings = () => {
  updateHttpsOnly();
  updateDefaultSearch();
};

// Update browser extensions recommendations
const getCurrentUserRecommendations = () => {
  updateExtensions();
  updateSettings();
};

// Once recommendations has been returned from localStorage,
// we need to update to match what the user has installed
const stop = watch(recommendations, () => {
  getCurrentUserRecommendations();
  stop();
});

const useRecommendations = () => {
  const activeRecommendations = computed(() =>
    recommendations.value.filter((rec) => !rec.activated && !rec.ignored),
  );

  return {
    recommendations,
    activeRecommendations,
    updateRecommendation,
    getRecommendationById,
    cleanOutdated,
  };
};

export default useRecommendations;
