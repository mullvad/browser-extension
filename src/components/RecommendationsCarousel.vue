<script lang="ts" setup>
import { NCard, NCarousel, NIcon, NImage } from 'naive-ui';

import { closePopup } from '@/helpers/closePopup';

import Button from '@/components/Buttons/Button.vue';
import MdiArrowLeft from '@/components/Icons/MdiArrowLeft.vue';
import MdiArrowRight from '@/components/Icons/MdiArrowRight.vue';
import IconLabel from '@/components/IconLabel.vue';

import useRecommendations from '@/composables/useRecommendations/useRecommendations';

const { activeRecommendations } = useRecommendations();
</script>

<template>
  <h1 class="text-sm pb-1">Privacy recommendations</h1>

  <div v-if="activeRecommendations.length === 0">
    <p class="text-white text-lg" data-test="success-message">
      Sweet! You have taken action on all recommendations.
    </p>
    <router-link class="hover:text-white" to="privacy-recommendations">
      See all Privacy Recommendations
    </router-link>
  </div>

  <div v-else>
    <n-carousel show-arrow>
      <template #arrow="{ prev, next }">
        <div class="custom-arrow">
          <n-icon class="arrow-icon" size="25" @click="prev"><MdiArrowLeft /></n-icon>
          <n-icon class="arrow-icon" size="25" @click="next"><MdiArrowRight /></n-icon>
        </div>
      </template>

      <n-card
        v-for="(recommendation, index) in activeRecommendations"
        :key="index"
        :bordered="false"
      >
        <template #header>
          <div class="flex">
            <n-image
              v-if="recommendation.icon"
              class="mr-4"
              width="20"
              :src="`/assets/icons/${recommendation.icon}`"
              object-fit="contain"
              preview-disabled
            />
            <h3>{{ recommendation.name }}</h3>
          </div>
        </template>
        <template #header-extra>
          <p>{{ index + 1 }} / {{ activeRecommendations.length }}</p>
        </template>

        <p>
          {{ recommendation.description }}
        </p>

        <template #footer>
          <Button
            v-if="recommendation.ctaUrl"
            :href="recommendation.ctaUrl"
            class="mr-4"
            @click="closePopup"
          >
            <IconLabel
              v-if="recommendation.ctaLabel"
              type="external"
              :text="recommendation.ctaLabel"
            />
          </Button>
          <router-link
            :to="`/privacy-recommendations#${recommendation.id}`"
            class="hover:text-white underline"
          >
            Learn more
          </router-link>
        </template>
      </n-card>
    </n-carousel>

    <div class="text-right pt-2 mr-2">
      <router-link class="hover:text-white" to="privacy-recommendations" data-test="show-all-link"
        >Show all</router-link
      >
    </div>
  </div>
</template>

<style>
.custom-arrow {
  display: flex;
  position: absolute;
  align-items: center;
  bottom: 9px;
  right: 11px;
}

.custom-arrow .n-icon {
  margin-left: 5px;
  cursor: pointer;
}

.custom-arrow .n-icon:hover {
  color: white;
}

.custom-arrow .n-icon:active {
  transform: scale(0.95);
  transform-origin: center;
}

.n-carousel.n-carousel--bottom .n-carousel__dots {
  display: none !important;
}
</style>
