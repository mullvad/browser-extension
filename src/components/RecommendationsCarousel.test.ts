import { RouterLinkStub, mount } from '@vue/test-utils';

import RecommendationsCarousel from '@/components/RecommendationsCarousel.vue';

import useRecommendations from '@/composables/useRecommendations/useRecommendations';

jest.mock('@/composables/useRecommendations/useRecommendations', () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe('RecommendationsCarousel', () => {
  it('should show a success message when no more recommendations are left', () => {
    (useRecommendations as jest.Mock).mockReturnValueOnce({ activeRecommendations: [] });
    const wrapper = mount(RecommendationsCarousel, {
      global: { stubs: { RouterLink: RouterLinkStub } },
    });

    const spanTag = wrapper.find('[data-test="success-message"]');
    expect(spanTag.exists()).toBe(true);
    // the reg exp flag "i" means case-insensitive, so you can ignore casing in the test
    expect(spanTag.text()).toMatch(/sweet! you have taken action on all recommendations.$/i);
    expect(wrapper.element).toMatchSnapshot();
  });
});
