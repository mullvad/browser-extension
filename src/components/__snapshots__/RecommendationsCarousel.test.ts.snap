// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecommendationsCarousel should show a "Show all" link when there are recommendations 1`] = `
<div
  data-v-app=""
>
  
  <title-category-stub
    class="my-2"
    title="Privacy recommendations"
  />
  <div
    class="-mb-1"
  >
    <carousel-stub
      autoplay="false"
      centeredslides="false"
      defaultindex="0"
      direction="horizontal"
      dotplacement="bottom"
      dottype="dot"
      draggable="false"
      effect="slide"
      interval="5000"
      keyboard="false"
      loop="true"
      mousewheel="false"
      showarrow="true"
      showdots="false"
      slidesperview="1"
      spacebetween="0"
      touchable="true"
      transitionstyle="[object Object]"
      trigger="click"
    />
    <div
      class="text-right pt-2 mr-2"
    >
      <a
        class="hover:text-white"
        data-test="show-all-link"
      >
        Show all
      </a>
    </div>
  </div>
  
</div>
`;

exports[`RecommendationsCarousel should show a success message when no more recommendations are left 1`] = `
<div
  data-v-app=""
>
  
  <title-category-stub
    class="my-2"
    title="Privacy recommendations"
  />
  <div>
    <p
      class="text-white text-lg"
      data-test="success-message"
    >
       Sweet! You have taken action on all recommendations. 
    </p>
    <a
      class="hover:text-white"
    >
       See all Privacy Recommendations 
    </a>
  </div>
  
</div>
`;
